### PR TITLE
sql/distsqlrun: refactor mergeJoiner to implement RowSource

### PR DESCRIPTION
**Pull request by petermattis**
The performance diff at smaller input sizes is due to the removal of the
`log.VEventf` lines.

```
name                           old time/op    new time/op    delta
MergeJoiner/InputSize=0-8        3.39µs ± 2%    2.30µs ± 2%  -32.22%  (p=0.000 n=10+9)
MergeJoiner/InputSize=4-8        5.66µs ± 1%    4.19µs ± 1%  -25.91%  (p=0.000 n=10+8)
MergeJoiner/InputSize=16-8       8.76µs ± 1%    7.14µs ± 1%  -18.55%  (p=0.000 n=9+10)
MergeJoiner/InputSize=256-8      66.6µs ± 1%    64.3µs ± 1%   -3.50%  (p=0.000 n=9+9)
MergeJoiner/InputSize=4096-8      987µs ± 1%     991µs ± 1%     ~     (p=0.089 n=10+10)
MergeJoiner/InputSize=65536-8    16.1ms ± 3%    16.4ms ± 3%     ~     (p=0.113 n=9+10)
```

Release note: None

### CODE REVIEW

**Comment by petermattis** [comment]
I broke this change into a series of smaller commits so that the reviewer can see the transformations I made before arriving at the final state.

**Comment by petermattis** [comment]
> The performance diff at smaller input sizes is due to the removal of the log.VEventf lines.

This actually isn't the cause of the performance delta. It is something to do with the last commit. I'm investigating.

**Comment by petermattis** [comment]
The difference is actually due to not using `DrainAndClose`. That method spawns a goroutine if there is more than one input to drain:

```
		for _, input := range srcs[1:] {
			wg.Add(1)
			go func(input RowSource) {
				DrainAndForwardMetadata(ctx, input, dst)
				wg.Done()
			}(input)
		}
```

The refactored `mergeJoiner` drains inputs serially. I'm mildly surprised this makes a difference, but there you go.

File: pkg/sql/distsqlrun/mergejoiner.go
Comment (arjunravinarayan) [inline pkg/sql/distsqlrun/mergejoiner.go]: Elegant! Thanks for the comment. Would be good to get one more review on the refactored merge joiner state machine. @RaduBerinde, do you mind taking a quick look at this function?

**Review by rjnn — CHANGES_REQUESTED** [review]
FWIW I found it easier to review it by viewing all the diffs at once, although I do appreciate that you gave me the option of doing it both ways. Reviewable also borked out on me, so doing this the old fashioned way.

The stream group accumulator and stream merger changes LGTM. Let's get another set of eyes on the merge joiner state machine, and a resolution on the double `ConsumerDone`.

File: pkg/sql/distsqlrun/mergejoiner.go
Comment (petermattis) [reply pkg/sql/distsqlrun/mergejoiner.go]: Copy pasta. Fixed.

File: pkg/sql/distsqlrun/stream_group_accumulator.go
Comment (petermattis) [reply pkg/sql/distsqlrun/stream_group_accumulator.go]: Removed that portion of the sentence.

**Review by petermattis — COMMENTED** [review]
TFTR!

**Review by RaduBerinde — APPROVED** [review]
Thanks for breaking it up into commits! LGTM!

**Comment by petermattis** [comment]
> This should probably be a FastIntSet (though I don't think that should be part of this PR, just pointing it out)

Heh, I noticed that too. Do you have any sense for how big typical merge-join batches are?

@arjunravinarayan you need to acknowledge the changes you requested.

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21219#-:-L26y4n2V2QNuAQxNaot:b-twpqtg)*

**Comment by arjunravinarayan** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1, 4 of 4 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 3 of 3 files at r7, 3 of 4 files at r9, 3 of 3 files at r14.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21219#-:-L26yzyqag2-WEIp4U_7:bnfp4nl)*

**Comment by petermattis** [comment]
Hmm, I guess github wants something else (see the red X next to @arjunravinarayan above), but it is allowing me to merge, so away we go.

**Comment by arjunravinarayan** [comment]
I figured out what the problem was: reviewable borked out on me last night and I used github reviews, and I "requested changes", which put a big X. I needed to "approve" them in GitHub, not reviewable.

**Comment by RaduBerinde** [comment]
*[pkg/sql/distsqlrun/mergejoiner.go, line 167 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/21219#-L1yhb4Z_L3bGrHloPE0-r7-167:-L271N5vgd_Apr5GBzyL:b-p61ik9) ([raw file](https://github.com/cockroachdb/cockroach/blob/914565af6dbe75112928aafd87fb18d12c214f19/pkg/sql/distsqlrun/mergejoiner.go#L167)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Elegant! Thanks for the comment. Would be good to get one more review on the refactored merge joiner state machine. @RaduBerinde, do you mind taking a quick look at this function?
</blockquote></details>

It looks good to me. This part is easier to review with the separate commits.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21219)*

**Comment by RaduBerinde** [comment]
It's hard to give a blanket statement, but it's common for the merge-join batches to be small (e.g. 1-row batches when joining on a key).

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21219#-:-L271aOi4YMub-KAZPsc:b5bt9cz)*

**Pull request merged**
